[tohlh] iP

tohlh opened the pull request:
# Henry

> "Discipline is the bridge between goals and accomplishment." - Jim Rohn ([source](https://www.wanderlustworker.com/discipline-is-the-bridge-between-goals-and-accomplishment/#:~:text=It%20was%20Jim%20Rohn%20who,And%20that%27s%20no%20understatement.))

Henry is a _fancy_ chatbot that helps you manage your tasks, **giving you more _freedom_**.

Henry is:
- Text based
- Easy to use
- Fast

To use Henry:
1. Download it from [here](https://github.com/tohlh/ip)
2. Run `./gradlew run build` from the root directory
3. Open the jar file from `build/libs/henry.jar`
4. Add some tasks

And you are set! 🎉

Features:
- [x] Add, edit, and delete tasks
- [x] Sort tasks
- [x] Find tasks
- [ ] Reminders (coming soon)
- [x] GUI (coming soon)

You can use Duke in your project, example `main` method:
```java
public static void main(String[] args) {
    new Henry("data/tasks.txt").run();
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Commit:
docs/README.md: Tweak document template

Let's tweak the docs/README.md (which is used as the user guide)
to fit Duke better. Specifically,

1. mention product name in the title
2. mention adding a product screenshot and a product intro
3. tweak the flow to describe feature-by-feature

tohlh committed:
Rename, greet and exit

tohlh committed:
Echo

tohlh committed:
Add and list items

tohlh committed:
Add Task class

tohlh committed:
Mark and unmark Tasks

tohlh committed:
Add Todo, Event and Deadline classes

tohlh committed:
Use Todo, Event and Deadline

tohlh committed:
Automated Text UI Testing

tohlh committed:
Refactor into methods

tohlh committed:
Add HenryException

tohlh committed:
Error handling

tohlh committed:
Better command handling

tohlh committed:
Delete items

tohlh committed:
Use enums

tohlh committed:
Add saving Tasks to file

tohlh committed:
Merge branch 'branch-Level-7'

tohlh committed:
Use LocalDateTime for Deadline and Event

tohlh committed:
Merge branch 'branch-Level-8'

tohlh committed:
Refactor Ui into its own class

tohlh committed:
Refactor Storage into its own class

tohlh committed:
Refactor TaskList into its own class

tohlh committed:
Refactor ReadCommand into Ui

tohlh committed:
Refactor Commands into classes

tohlh committed:
Refactor Parser into its own class

tohlh committed:
Fix Event string representation

tohlh committed:
Merge branch 'branch-A-MoreOOP'

tohlh committed:
Refactor into packages

tohlh committed:
Merge branch 'branch-A-Packages'

tohlh committed:
Merge branch 'add-gradle-support'

tohlh committed:
Update build.gradle mainClass name

tohlh committed:
Improve parsing for Deadline and Event commands

tohlh committed:
Improve parsing number for Mark, Unmark and Delete

tohlh committed:
Add ParserTest

tohlh committed:
Add TodoTest

tohlh committed:
Change archiveBaseName

tohlh committed:
Add JavaDocs to all files

tohlh committed:
Tweak code to follow coding standard

tohlh committed:
Add Find tasks functionality

tohlh committed:
Merge branch 'branch-Level-9'

tohlh committed:
Merge branch 'branch-A-JavaDoc'

tohlh committed:
Add JavaDoc for findTasksByKeyword

tohlh committed:
Merge branch 'branch-A-CodingStandard'

tohlh committed:
Update gitignore

tohlh committed:
Configure CheckStyle and fix styles

tohlh committed:
Merge branch 'branch-A-CheckStyle'

tohlh committed:
Setup JavaFX

tohlh committed:
Refactor command to return string and remove UI class

tohlh committed:
Fix bye command

tohlh committed:
Fix code styles

tohlh committed:
Merge branch 'branch-Level-10'

doyelee0313 commented on src/main/java/henry/command/Command.java:
```diff
@@ -0,0 +1,28 @@
+/**
+ * Represents a command object that can be executed
+ */
+public abstract class Command {
```
This separation of commands and using abstract class is good

doyelee0313 commented on src/main/java/henry/ui/MainWindow.java:
```diff
@@ -0,0 +1,52 @@
+
+    private Henry henry;
+
+    private Image userImage = new Image(this.getClass().getResourceAsStream("/images/DaUser.png"));
```
Use of pictures for better GUI is good here. It is better than having just text UI.

doyelee0313 commented on src/main/java/henry/TaskList.java:
```diff
@@ -0,0 +1,117 @@
+     */
+    public String deleteTask(int index) throws HenryException {
+        if (index < 0 || index >= items.size()) {
+            throw new HenryException("The index is out of bounds!");
```
I like how you also handle the out of bound exception.

doyelee0313 commented on src/main/java/henry/ui/DialogBox.java:
```diff
@@ -0,0 +1,61 @@
+    }
+
+    public static DialogBox getDukeDialog(String text, Image img) {
+        var db = new DialogBox(text, img);
```
Variable name db may not be so specific but should be fine since it is only used in here.

dabzpengu commented on src/main/java/henry/TaskList.java:
```diff
@@ -0,0 +1,117 @@
+     * @param keyword The keyword to be searched for.
+     * @return The list of tasks that contain the keyword.
+     */
+    public ArrayList<Task> findTasksByKeyword(String keyword) {
```
i like how your method names are clear and meaningful

dabzpengu commented on src/main/java/henry/task/Task.java:
```diff
@@ -0,0 +1,73 @@
+     * @return The string to be stored in a file.
+     */
+    public String toFileString() {
+        return String.format("%s | %s", isDone ? "1" : "0", description);
```
I like how clean and concise this line is

dabzpengu commented on src/main/java/henry/command/DeleteCommand.java:
```diff
@@ -0,0 +1,36 @@
+
+    @Override
+    public String execute(TaskList tasks, Storage storage) throws HenryException {
+        String ret = tasks.deleteTask(index);
```
I think that the String variable name "ret" can be more meaningful

tohlh committed:
Change variable name for clarity

tohlh committed:
Add assertions in Parser

Parser splits a full command strings into command type and arguments. Some commands do not accept arguments but some do.

Assertions can help make explicit which commands require arguments and which do not.

Let's use assertions to document such assumptions in the future.

tohlh committed:
Make EventCommand constructor shorter

tohlh committed:
Fix code styles in a few files

tohlh committed:
Merge pull request #1 from tohlh/branch-A-Assertions

Add assertions in Parser

tohlh committed:
Merge pull request #2 from tohlh/branch-A-CodeQuality

Improve Code Quality

tohlh committed:
Create gradle.yml

tohlh committed:
Fix code styles and remove unused junit test

tohlh committed:
Add UpdateCommand

Adds an Update command for updating details of tasks added into the list.

Update command accepts an index, /d, /by, /from, /to parameters.

Other than index and /d, the other parameters are optional.

Let's improve the command in the future.

tohlh committed:
Merge branch 'branch-BCD-Extension'

tohlh committed:
Fix code styles

tohlh committed:
Greets the user on startup

tohlh committed:
Update README.md and add Ui.png

tohlh committed:
Add Delete feature into README.md

tohlh committed:
Update README.md

tohlh committed:
Update README.md formatting

tohlh committed:
Add bye command into README.md

tohlh committed:
Add tests for Deadline and Event

tohlh committed:
Fix typo in README.md

tohlh committed:
Add screenshot and example
